File: examples/C/set_pv.c
/**
* Copyright 2005-2007 ECMWF
* 
* Licensed under the GNU Lesser General Public License which
* incorporates the terms and conditions of version 3 of the GNU
* General Public License.
* See LICENSE and gpl-3.0.txt for details.
*/

/*
 * C Implementation: set
 *
 * Description: how to set pv .
 *
 *
 * Author: Enrico Fucile <enrico.fucile@ecmwf.int>
 *
 *
 */
#include <stdio.h>
#include <stdlib.h>

#include "grib_api.h"

int main(int argc, char** argv) {
  int err = 0;
  long PVPresent=1;
  size_t size=0;
  double pv[4]={1,2,3,4};
  size_t pvsize=4;

  FILE* in = NULL;
  char* infile = "../data/regular_latlon_surface.grib1";
  FILE* out = NULL;
  char* outfile = "out.grib1";
  grib_handle *h = NULL;
  const void* buffer = NULL;

  in = fopen(infile,"r");
  if(!in) {
    printf("ERROR: unable to open file %s\n",infile);
    return 1;
  }

  out = fopen(outfile,"w");
  if(!in) {
    printf("ERROR: unable to open file %s\n",outfile);
    return 1;
  }

  /* create a new handle from a message in a file */
  h = grib_handle_new_from_file(0,in,&err);
  if (h == NULL) {
    printf("Error: unable to create handle from file %s\n",infile);
  }

  GRIB_CHECK(grib_set_long(h,"PVPresent",PVPresent),0);


  GRIB_CHECK(grib_set_double_array(h,"pv",pv,pvsize),0);

  /* get the coded message in a buffer */
  GRIB_CHECK(grib_get_message(h,&buffer,&size),0);

  /* write the buffer in a file*/
  if(fwrite(buffer,1,size,out) != size) 
  {
     perror(argv[1]);
     exit(1);
  }

  /* delete handle */
  grib_handle_delete(h);

  fclose(in);
  fclose(out);

  return 0;
}

File: examples/F90/set_pv.f90
! Copyright 2005-2007 ECMWF
! 
! Licensed under the GNU Lesser General Public License which
! incorporates the terms and conditions of version 3 of the GNU
! General Public License.
! See LICENSE and gpl-3.0.txt for details.
!
!
!  Description: how to set pv values.
!
!
!  Author: Anne Fouilloux
!
!
program set_pv
  use grib_api
  implicit none
  integer                         :: numberOfLevels
  integer                         :: numberOfCoefficients
  integer                         :: outfile, igrib
  integer                         :: i, ios
  real, dimension(:),allocatable  :: pv
  
  numberOfLevels=60
  numberOfCoefficients=2*(numberOfLevels+1)

  allocate(pv(numberOfCoefficients))

  ! read the model level coefficients from file
  open( unit=1, file="../../data/60_model_levels", &
                form="formatted",action="read")

  do i=1,numberOfCoefficients,2
     read(unit=1,fmt=*, iostat=ios) pv(i), pv(i+1)
     if (ios /= 0) then
        print *, "I/O error: ",ios
        exit
     end if
  end do
  
  ! print coefficients
  !do i=1,numberOfCoefficients,2
  !  print *,"  a=",pv(i)," b=",pv(i+1)
  !end do

  close(unit=1)

  call grib_open_file(outfile, 'out.grib1','w')
  
  !     a new grib message is loaded from file
  !     igrib is the grib id to be used in subsequent calls
  call grib_new_from_samples(igrib, "reduced_gg_sfc_grib1")

  !     set levtype to ml (model level)
  call grib_set(igrib,'typeOfLevel','hybrid')

  !     set level 
  call grib_set(igrib,'level',2)

  !     set PVPresent as an integer 
  call grib_set(igrib,'PVPresent',1)
  
  call grib_set(igrib,'pv',pv)
  
  !     write modified message to a file
  call grib_write(igrib,outfile)
  
  !  FREE MEMORY
  call grib_release(igrib)
  deallocate(pv)

  call grib_close_file(outfile)
  
end program set_pv
